[`pylint`] Implement rule to prefer augmented assignment (`PLR6104`)

lshi18 opened the pull request:
<!--
Thank you for contributing to Ruff! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title?
- Does this pull request include references to any relevant issues?
-->

## Summary

<!-- What's the purpose of the change? What does it do, and why? -->
Implement new rule: Prefer augmented assignment (#8877). It checks for the assignment statement with the form of `<expr> = <expr> <binary-operator> …` with a unsafe fix to use augmented assignment instead.

## Test Plan

<!-- How was it tested? -->
1. Snapshot test is included in the PR.
2. Manually test with playground.

Avasam commented:
Not a Rust dev, but I think the tests should check for handling of operation stuff as well. The following should not trigger RUF028:
```python
>>> test = 2
>>> test = test * test + 10
>>> test
14
>>> test = 2
>>> test *= test + 10       
>>> test
24
```

lshi18 commented:
Thanks for the quick reply. I added the suggested test cases. I had these two test cases in my earlier draft, but wasn’t sure and removed them. It is nice that you brought them up and have them back.

sbrugman commented:
Related pylint rule: 
https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/consider-using-augmented-assign.html

lshi18 commented:
The new rule has been moved under PLR6104 instead of RUF028.

lshi18 commented:
The `CI/ecosystem` check failed with an error. I am not able to figure out what I did wrong immediately. It'll be of great help if you could point me to the right direction which could effect an efficient fix.

sbrugman commented on crates/ruff_linter/src/rules/pylint/rules/binary_op_and_normal_assignment.rs:
```diff
@@ -0,0 +1,230 @@
+    checker: &mut Checker,
+    assign @ ast::StmtAssign { value, targets, .. }: &ast::StmtAssign,
+) {
+    if targets.len() != 1 {
```
Using [let-else](https://doc.rust-lang.org/beta/rust-by-example/flow_control/let_else.html) here to make this check a bit nicer:

```rust
let target = Some(targets.first()) else {
    return;
}
```

lshi18 replied:
I don't think they are semantically equal. In particular, the following test case which is marked as `OK` fails after the suggested change.

```python
a_number = index = a_number + 1 # OK
```

I think a further question would be whether the should detect the above case and, in the same gist, the one below.

```python
index = a_number = a_number + 1 # OK
```

lshi18 replied:
I had a quick test with `pylint 3.0.3`, whose `R6104` rule detects the 

```python
a_number = index = a_number + 1 # OK
```

but not this,
```python
index = a_number = a_number + 1 # OK
```

Do you think we should implement to reproduce this effect?

sbrugman commented on crates/ruff_linter/src/rules/pylint/rules/binary_op_and_normal_assignment.rs:
```diff
@@ -0,0 +1,230 @@
+    }
+    let target = targets.first().unwrap();
+
+    let rhs_expr = value
```
```rust
 let rhs_expr = Some(
      value
       .as_bin_op_expr()
       .map(|e| (e.left.as_ref(), e.op, e.right.as_ref()))
      ) else {
       return;
}
```

sbrugman commented on crates/ruff_linter/src/rules/pylint/rules/binary_op_and_normal_assignment.rs:
```diff
@@ -0,0 +1,230 @@
+    }
+    let (left_operand, operator, right_operand) = rhs_expr.unwrap();
+
+    if name_expr(target, left_operand)
```
Instead of testing these conditions, you could match the target/left_operand type with match.

There may be some other rules to serve as inspiration e.g.
https://github.com/astral-sh/ruff/blob/235cfb79769da2c435b9c88d8bae4a79f1234857/crates/ruff_linter/src/rules/flake8_bugbear/rules/static_key_dict_comprehension.rs#L73
